### PR TITLE
Validation Thread

### PR DESCRIPTION
**Pull request by filipepgm**
Creates a thread to perform validation in the master. Validations are submitted via a FIFO queue.
There is a separate tensorflow graph, session and model where the validation is run.
Activated by the flag "--thread_validation"

### CODE REVIEW

**Comment by vlimant** [comment]
can you please do the trivial rebasing for the fix on epoch ?

**Comment by vlimant** [comment]
have you verified that both keras+th and torch function well with this ? it looks a bit involved, is there any penalty of copying the model around ? do we end up with two of them on the GPU?

**Comment by filipepgm** [comment]
Rebase done.
I have run the standard mnist example with torch, tf and theano backends, all converging. Based on this, I assume that it is working well with all three backends.
My understanding is that both models will be available in the gpu, one of them only computing updates, and the other only running over the validation data.

For performance, I looked at the cifar10 example with tf backend.
In a normal execution, validation represents around 5% of execution time on the master. As such, I couldn't observe any advantage in using the thread.
In the trace with the thread, it is possible to observe that the master does keep on answering workers while validating, although slower.

Increasing artificially the rate of validation, so that it takes around half of the execution time, makes the use of the validation thread have some impact. Over a number of cifar10 + tensorflow executions, executions with no thread took around 100 seconds, and with thread, around 80 seconds.

The thread doesn't seem to have perfect parallelism. It also seems to only have a real impact when the validation is performed relatively often.

**Comment by vlimant** [comment]
looks good to me
